discovery/ecs/ecs.go

```diff
@@ -0,0 +1,150 @@
+package ecs
+
+import (
+	"net"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/ecs"
+	"github.com/docker/swarm/discovery"
+)
+
+// Discovery is exported.
+type Discovery struct {
+	ecs           ecsClient
+	ec2           ec2Client
+	cluster, port string
+	heartbeat     time.Duration
+}
+
+func init() {
+	Init()
+}
+
+// Init is exported.
+func Init() {
+	discovery.Register("ecs", &Discovery{})
+}
+
+func parsePath(path string) (cluster string, port string) {
+	cluster, port, err := net.SplitHostPort(path)
+	if err != nil {
+		return path, "2375"
+	}
+	return cluster, port
+}
+
+// Initialize is exported.
+func (s *Discovery) Initialize(path string, heartbeat time.Duration, ttl time.Duration) error {
+	config := aws.DefaultConfig
+	s.heartbeat = heartbeat
+	s.ecs = ecs.New(config)
+	s.ec2 = ec2.New(config)
+	s.cluster, s.port = parsePath(path)
+	return nil
+}
+
+func (s *Discovery) fetch() (discovery.Entries, error) {
+	var entries discovery.Entries
+
+	list, err := s.ecs.ListContainerInstances(&ecs.ListContainerInstancesInput{
+		Cluster: aws.String(s.cluster),
+	})
+	if err != nil {
+		return entries, err
+	}
+
+	ecsDesc, err := s.ecs.DescribeContainerInstances(&ecs.DescribeContainerInstancesInput{
+		Cluster:            aws.String(s.cluster),
+		ContainerInstances: list.ContainerInstanceARNs,
+	})
+	if err != nil {
+		return entries, err
+	}
+
+	var instances []*string
+	for _, instance := range ecsDesc.ContainerInstances {
+		instances = append(instances, instance.EC2InstanceID)
+	}
+
+	ec2Desc, err := s.ec2.DescribeInstances(&ec2.DescribeInstancesInput{
+		Filters: []*ec2.Filter{
+			&ec2.Filter{
+				Name:   aws.String("instance-id"),
+				Values: instances,
+			},
+		},
+	})
+	if err != nil {
+		return entries, err
+	}
+
+	for _, reservation := range ec2Desc.Reservations {
+		for _, instance := range reservation.Instances {
+			entries = append(entries, &discovery.Entry{
+				Host: *instance.PrivateIPAddress,
+				Port: s.port,
+			})
+		}
+	}
+
+	return entries, nil
+}
+
+// Watch is exported
+func (s *Discovery) Watch(stopCh <-chan struct{}) (<-chan discovery.Entries, <-chan error) {
```

Comment:
This is a direct clone of [file.Discovery#Watch](https://github.com/docker/swarm/blob/729001c9b458d7467986b189ac9b6647191a0fff/discovery/file/file.go#L64). I could extract a more generic Discovery that takes a `Fetch() (discovery.Entries, error)` interface. Dunno if it's really worth it.